### PR TITLE
Add Aborted conditionType for policy enactments

### PR DESCRIPTION
**Pull request by rhrazdil**
Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

**Is this a BUG FIX or a FEATURE ?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
/kind enhancement

**What this PR does / why we need it**:
When rolling out nncp sequentially, if a node fails to apply configuration
other nodes do not attempt to apply it and set their enactment condition to
Aborted.

Example:
We apply sequential `policy` in a 3 node cluster.
Assume the fist node that starts applying configuration is node01, then the result will look like:
```
$ kubectl get nnce
NAME            STATUS
node01.policy   FailedToConfigure
node02.policy   ConfigurationAborted
node03.policy   ConfigurationAborted
```
In other words, node01 has started applying and failed. Node02 and node03 noted that there is a node failing to apply `policy`
and instead of attempting to apply it too set their enactment status to `ConfigurationAborted`.

**Special notes for your reviewer**:

**Release note**:

```release-note
Added new conditionType Aborted used in sequential configuration rollout
```

### CODE REVIEW

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 1]
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
...
+func (r *NodeNetworkConfigurationPolicyReconciler) checkEnactmentConditions(policy *nmstatev1beta1.NodeNetworkConfigurationPolicy) error {
+	enactments := nmstatev1beta1.NodeNetworkConfigurationEnactmentList{}
+	policyLabelFilter := client.MatchingLabels{nmstateapi.EnactmentPolicyLabel: policy.Name}
+	err := r.Client.List(context.TODO(), &enactments, policyLabelFilter)
+	if err != nil {
+		return err
+	}
+	enactmentsCount := enactmentconditions.Count(enactments, policy.Generation)
+	if enactmentsCount.Failed() > 0 {
+		// a node has failed to apply configuration
```
Comment (qinqon) [inline controllers/nodenetworkconfigurationpolicy_controller.go:156]: Comment not needed since the error explain it.

File: pkg/policyconditions/conditions.go [hunk 2]
```diff
@@ -152,8 +152,11 @@ func Update(cli client.Client, policyKey types.NamespacedName) error {
 			if enactmentsCount.Matching() == 0 {
 				message := "Policy does not match any node"
 				setPolicyNotMatching(&policy.Status.Conditions, message)
-			} else if enactmentsCount.Failed() > 0 {
+			} else if enactmentsCount.Failed() > 0 || enactmentsCount.Aborted() > 0 {
 				message := fmt.Sprintf("%d/%d nodes failed to configure", enactmentsCount.Failed(), enactmentsCount.Matching())
+				if enactmentsCount.Aborted() > 0 {
+					message += fmt.Sprintf(", %d nodes aborted configuration", enactmentsCount.Aborted())
```
Comment (qinqon) [inline pkg/policyconditions/conditions.go:158]: Let's add also the number of matching

**Review by qinqon — CHANGES_REQUESTED** [review]
first pass: Skipping review of tests
some nits and a question
but looks pretty good.

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 3]
```diff
@@ -161,19 +183,6 @@ func (r *NodeNetworkConfigurationPolicyReconciler) setNodeRunningUpdate(policyKe
...
 
-func (r *NodeNetworkConfigurationPolicyReconciler) claimNodeRunningUpdate(
-	policy *nmstatev1beta1.NodeNetworkConfigurationPolicy,
-	ec *enactmentconditions.EnactmentConditions,
-) error {
-	if policy.Status.NodeRunningUpdate != "" && metav1.Now().Sub(policy.Status.NodeUpdateStart.Time) > nodeConfigurationTimeout {
-		// a node has been running the update for too long
-		err := fmt.Errorf("A node has been configuring for too long, aborting")
-		ec.NotifyFailedToConfigure(err)
-		return err
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go:172]: I only moved it to another function called before `claimNodeRunningUpdate`. I didn't like that function that is named `claimSomething` does a lot of other things apart of claimingSomething.

File: pkg/policyconditions/conditions.go (hunk 2 above)
Comment (rhrazdil) [reply pkg/policyconditions/conditions.go:158]: it is still there, I concatenate it to the original `message`

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 4]
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
...
+
+func (r *NodeNetworkConfigurationPolicyReconciler) claimNodeRunningUpdate(policy *nmstatev1beta1.NodeNetworkConfigurationPolicy) error {
+	policyKey := types.NamespacedName{Name: policy.GetName(), Namespace: policy.GetNamespace()}
 	err := r.Client.Get(context.TODO(), policyKey, policy)
 	if err != nil {
 		return err
 	}
 	if policy.Status.NodeRunningUpdate != "" {
-		return fmt.Errorf("another node is working on configuration")
+		return apierrors.NewConflict(schema.GroupResource{Resource: "nodenetworkconfigurationpolicies"}, policy.Name, fmt.Errorf("Another node is working on configuration"))
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go:175]: Yes indeed

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 5]
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
...
+	if err != nil {
+		return err
+	}
+	enactmentsCount := enactmentconditions.Count(enactments, policy.Generation)
+	if enactmentsCount.Failed() > 0 {
+		// a node has failed to apply configuration
+		err := fmt.Errorf("current node has failed to apply policy configuration")
+		return err
+	}
+	if policy.Status.NodeRunningUpdate != "" && metav1.Now().Sub(policy.Status.NodeUpdateStart.Time) > nodeConfigurationTimeout {
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go:160]: That's a good point. It looks like we don't need the timestamp at all then

File: controllers/nodenetworkconfigurationpolicy_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: Done

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 6]
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
...
+	policyLabelFilter := client.MatchingLabels{nmstateapi.EnactmentPolicyLabel: policy.Name}
+	err := r.Client.List(context.TODO(), &enactments, policyLabelFilter)
+	if err != nil {
+		return err
+	}
+	enactmentsCount := enactmentconditions.Count(enactments, policy.Generation)
+	if enactmentsCount.Failed() > 0 {
+		// a node has failed to apply configuration
+		err := fmt.Errorf("current node has failed to apply policy configuration")
+		return err
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: Done

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 7]
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
...
+	}
+	if policy.Status.NodeRunningUpdate != "" && metav1.Now().Sub(policy.Status.NodeUpdateStart.Time) > nodeConfigurationTimeout {
+		// a node has been running the update for too long
+		err := fmt.Errorf("current node has been configuring for too long")
+		return err
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: Done

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 8]
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
...
+	enactments := nmstatev1beta1.NodeNetworkConfigurationEnactmentList{}
+	policyLabelFilter := client.MatchingLabels{nmstateapi.EnactmentPolicyLabel: policy.Name}
+	err := r.Client.List(context.TODO(), &enactments, policyLabelFilter)
+	if err != nil {
+		return err
+	}
+	enactmentsCount := enactmentconditions.Count(enactments, policy.Generation)
+	if enactmentsCount.Failed() > 0 {
+		// a node has failed to apply configuration
+		err := fmt.Errorf("current node has failed to apply policy configuration")
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: Right, done

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 9]
```diff
@@ -55,7 +56,6 @@ import (
 
 var (
 	nodeName                                string
-	nodeConfigurationTimeout                = 5 * time.Minute
```
Comment (qinqon) [inline controllers/nodenetworkconfigurationpolicy_controller.go:58]: We don't need this ?

**Review by qinqon — CHANGES_REQUESTED** [review]
We don't need the configuration timeout anymore ? 

Also looks like docs/assets/images/logo/fullcolor.png has being overwritten.

File: controllers/nodenetworkconfigurationpolicy_controller.go (hunk 9 above)
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go:58]: Since we don't need to check how long a node is applying the policy, I think we can remove it, it's not being used except for setting it

Comment (qinqon) [reply controllers/nodenetworkconfigurationpolicy_controller.go:58]: You mean that we already have a timeout mechanism when we run `nmstatectl set` so we don't need any of this right ?

Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go:58]: Yes, I believe it's enough to rely on the timeout of `ApplyDesiredState` and only check if there is a node with Failed enactment

**Comment by rhrazdil** [comment]
> Also looks like docs/assets/images/logo/fullcolor.png has being overwritten.

you're right, thanks. Removed this change

**Comment by rhrazdil** [comment]
in the last push I have
- removed the timestamp of claiming nodeRunningUpdate
- fixed the test that failed in parallel lane, it wasn't expecting the aborted state to be there

File: pkg/enactmentstatus/conditions/counter_test.go [hunk 10]
```diff
@@ -154,6 +162,35 @@ var _ = Describe("Enactment condition counter", func() {
...
+			),
+			expectedCount: ConditionCount{
+				available:   CountByConditionStatus{t: 0, f: 2, u: 0},
+				failing:     CountByConditionStatus{t: 1, f: 1, u: 0},
+				progressing: CountByConditionStatus{t: 0, f: 2, u: 0},
+				matching:    CountByConditionStatus{t: 2, f: 0, u: 0},
+				aborted:     CountByConditionStatus{t: 1, f: 1, u: 0},
+			},
+		}),
+		Entry("e(Matching, Progressing), e(Matching, Aborted)", EnactmentCounterCase{
```
Comment (qinqon) [inline pkg/enactmentstatus/conditions/counter_test.go:182]: Cannot be progressing + aborted I think we can remove this test.

File: test/e2e/handler/nnce_conditions_test.go [hunk 11]
```diff
@@ -110,11 +141,33 @@ var _ = Describe("[rfe_id:3503][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
+			By("Check policy is at degraded state")
+			waitForDegradedTestPolicy()
+
+			By("Check that the enactment stays in failing state")
+			var wg sync.WaitGroup
+			wg.Add(len(nodes))
+			for i, _ := range nodes {
+				node := nodes[i]
+				go func() {
+					defer wg.Done()
```
Comment (qinqon) [inline test/e2e/handler/nnce_conditions_test.go:158]: we are missing the `defer GinkgoRecover`

File: test/e2e/handler/nnce_conditions_test.go [hunk 12]
```diff
@@ -128,23 +181,36 @@ var _ = Describe("[rfe_id:3503][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
+			for _, node := range nodes {
+				By(fmt.Sprintf("Check %s failing or aborted state is reached", node))
+				enactmentConditionsStatusEventually(node).Should(
+					SatisfyAny(
+						ConsistOf(failingEnactmentConditions...),
+						ConsistOf(abortedEnactmentConditions...),
+					), "should eventually reach failing of aborted conditions at enactments")
+			}
+
+			By("Check that the enactment stays in failing or aborted state")
```
Comment (qinqon) [inline test/e2e/handler/nnce_conditions_test.go:201]: This will pass is something is broken and we have two failures and one aborted at sequencial rollout we will have just one failed enactment.

File: test/e2e/handler/nnce_conditions_test.go [hunk 13]
```diff
@@ -128,23 +181,36 @@ var _ = Describe("[rfe_id:3503][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
 				node := nodes[i]
 				go func() {
 					defer wg.Done()
 					By(fmt.Sprintf("Check %s failing state is kept", node))
-					enactmentConditionsStatusConsistently(node).Should(ConsistOf(failingEnactmentConditions...), "should consistently keep failing conditions at enactments")
+					enactmentConditionsStatusConsistently(node).Should(
+						SatisfyAny(
+							ConsistOf(failingEnactmentConditions...),
+							ConsistOf(abortedEnactmentConditions...),
+						), "should consistently keep failing or aborted conditions at enactments")
```
Comment (qinqon) [inline test/e2e/handler/nnce_conditions_test.go:213]: ditto

**Review by qinqon — CHANGES_REQUESTED** [review]
Looking even better, I see code shrinking :-), e2e test reviewed too.

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 14]
```diff
@@ -166,48 +165,45 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
...
+	err := r.Client.List(context.TODO(), &enactments, policyLabelFilter)
+	if err != nil {
+		return err
+	}
+	enactmentsCount := enactmentconditions.Count(enactments, policy.Generation)
+	if enactmentsCount.Failed() > 0 {
+		return fmt.Errorf("some node/s has failed to apply policy configuration")
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: Done

File: controllers/nodenetworkconfigurationpolicy_controller.go [hunk 15]
```diff
@@ -263,9 +259,15 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(request ctrl.Reques
...
 	if !instance.Spec.Parallel {
-		err := r.claimNodeRunningUpdate(instance, &enactmentConditions)
+		err = r.checkForFailedEnactments(instance)
+		if err != nil {
+			// some enactments are failing, abort and don't requeue
+			log.Error(err, "policy has failing enactments, aborting")
+			enactmentConditions.NotifyAborted(err)
+			return ctrl.Result{}, nil
+		}
+		err = r.claimNodeRunningUpdate(instance)
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply controllers/nodenetworkconfigurationpolicy_controller.go]: Makes sense, thanks. Done

File: pkg/enactmentstatus/conditions/counter_test.go (hunk 10 above)
Comment (rhrazdil) [reply pkg/enactmentstatus/conditions/counter_test.go]: Done

File: test/e2e/handler/nnce_conditions_test.go (hunk 11 above)
Comment (rhrazdil) [reply test/e2e/handler/nnce_conditions_test.go:157]: Added

File: test/e2e/handler/nnce_conditions_test.go (hunk 12 above)
Comment (rhrazdil) [reply test/e2e/handler/nnce_conditions_test.go]: Addressed by adding a function that counts failing and aborted enactments and expects 1 failing and len(nodes)-1 aborted enactments

File: test/e2e/handler/nnce_conditions_test.go (hunk 13 above)
Comment (rhrazdil) [reply test/e2e/handler/nnce_conditions_test.go:226]: I think in this case, we can keep the or, because should any of those enactment change, it would need to be at least in `matching` or `notMatching`.
Afterwards I again check the numbers of aborted and failing
